Update the Asset Pipeline Guide [skip ci] (#45400)
rails/rails: guides/source/asset_pipeline.md

The Asset Pipeline
==================

> This guide covers the asset pipeline as it was implemented by Sprockets.

Reviewer: Doesn't this describe the Import Maps and Sprockets guide?
```suggestion
This guide covers the asset pipeline using Import Maps and Sprockets.
```

Maybe also link to the Sprockets guide:
```
The Sprockets only asset pipeline guide has been moved to a seperate guide.
```

* How to decide which approach to use in your own application;
* Where to go to read the guide for your chosen approach.

--------------------------------------------------------------------------------

What is the Asset Pipeline?
---------------------------

The asset pipeline is responsible for serving javascript, css and image files
in the most efficient manner possible, from your app, its ruby gems and its node
packages.

### Main Features

### What is Fingerprinting and Why Should I Care?

### What is different between delivering assets in HTTP1 and HTTP2?

The Four Approaches to the Asset Pipeline
-----------------------------------------

### Sprockets
The original asset pipeline gem, built for the HTTP/1 era and low javascript frontends. It handled the bundling and digesting of javascript, css and image files, without relying on node packages.

### Webpacker
Shipped with Rails 5.2, as a wrapper around the complexity of Webpack/Node/Yarn, this gem could completely replace Sprockets or simply take over javascript transpiling and bundling. It provided Rails “out of the box” support for SPA frameworks like React.

### Import Maps

> Shipped with Rails 7.0, it replaces Sprockets as the default asset pipeline gem. Although it eliminates the need for node/yarn and other complex tooling, it requires the application using it to be deployed in an environment that supports HTTP/2, otherwise it causes severe performance problems.

Reviewer 2: It probably should be meantioned that import maps are only for Javascript and not CSS? That's been a big gotcha I've noticed from people asking about it. They are expecting import maps to also handle the CSS.

Reviewer 3: I'd go one step further and tell people what they should use for CSS. Saying that importmaps is just for JS still confuses me when I try to make sense of Rails 7+ assets. The JS story is clear to me, but CSS remains unclear.

Reviewer 4: Just wait until module type assertions come to the browser and then this is no longer true...

🙈 

https://web.dev/css-module-scripts/
